src/snagrecover/protocols/imx_sdp.py

```diff
@@ -145,49 +147,157 @@ def write32(self, addr: int, value: int) -> bool:
 		logger.debug(f"Sending SDP packet {packet}")
 		self.dev.write(packet)
 		self.check_hab()
-		complete_status = self.dev.read(65, timeout=5000)[1:5]
+		complete_status = self._read(64, timeout=5000)[:4]
 		self.end_cmd()
 		return complete_status == b"\x12\x8A\x8A\x12"
 
 	def write_dcd(self, blob: bytes, addr: int, offset: int, size: int) -> bool:
-		return self.write_blob(blob, addr, offset, size, write_dcd=True)
+		if self.is_hid():
+			return self.write_blob(blob, addr, offset, size, write_dcd=True)
+
+		# Non HID devices do not have a DCD_WRITE command
+		# They do have a "FileType" byte in WRITE_FILE but that does not
+		# appear to actually process the DCD immediately (which we need to
+		# configure dram to download u-boot)
+		# So instead we manually interpret the DCD as individual read / write
+		# operations. This is what imx_usb_loader does too.
+		tag, lg, version = struct.unpack(">BHB", blob[offset:offset + 4])
+		if tag != 0xD2:
+			raise ValueError("Bad DCD tag %02x" % tag)
+		if version != 0x40:
+			raise ValueError("Bad DCD version %02x" % version)
+
+		pos = offset + 4
+		end = offset + lg
+		while pos < end:
+			tag, lg, param = struct.unpack(">BHB", blob[pos:pos + 4])
+
+			def invoke_for_each_addr_data(fn, param=param, pos=pos, lg=lg):
```

Comment:
I find this a bit hard to read. Could you please move this function definition outside of the SDPCommand class def? I think it would be more appropriate since it doesn't explicitly use the class attributes. Plus, it would tame the indentation a bit.